network/src/commonMain/kotlin/com/wire/kalium/network/utils/NetworkUtils.kt

```diff
@@ -121,54 +116,62 @@ internal fun <T : Any> NetworkResponse<T>.onFailure(fn: (NetworkResponse.Error)
-            is ServerResponseException -> {
-                // 500 .. 599
-                // TODO: do 500 errors have body
-                NetworkResponse.Error(kException = KaliumException.ServerError(e.response.body()))
+/**
+ * Wraps exceptions thrown during the request or parsing of the response, like
+ * exceptions thrown due to inability to reach the server.
+ * This does **not** handle the response itself (i.e. HTTP status, body, etc.),
+ * but may be used to catch exceptions thrown during those steps.
+ * 
+ * @return [NetworkResponse.Success] if everything went smooth
+ * @return [NetworkResponse.Error] with a [KaliumException.GenericError] in case of an error
+ */
+private inline fun <reified ResponseType: Any> handlingNetworkException(
+    performRequest: () -> NetworkResponse<ResponseType>
+): NetworkResponse<ResponseType> = try {
+    performRequest()
+} catch (e: Exception) {
+    NetworkResponse.Error(KaliumException.GenericError(e))
+}
+
+/**
+ * Wraps a producer of [HttpResponse] and attempts to parse the server response based on the [BodyType].
+ * @return - Successful response (HTTP Status Codes from 200 to 299):
+ * a [NetworkResponse.Success] with the expected [BodyType] will be returned.
+ *
+ * - Unsuccessful response (any other HTTP Status Code):
+ * a [NetworkResponse.Error] with a [KaliumException]. Will try to read it the body as an [ErrorResponse].
+ * It's possible to intercept this and do the mapping to a [NetworkResponse] through [unsuccessfulResponseOverride].
+ *
+ * - Exceptions failure to reach server or parse response:
+ * a [NetworkResponse.Error] containing a [KaliumException.GenericError]
+ *
+ * @param unsuccessfulResponseOverride Allows to intercept any unsuccessful response
+ * and map it to a [NetworkResponse] as needed. Useful when handling custom ErrorBody
+ * @param performRequest the block that will result into the [HttpResponse]
+ * @see KaliumException
+ * @see ErrorResponse
+ */
+internal suspend inline fun <reified BodyType : Any> wrapKaliumResponse(
+    unsuccessfulResponseOverride: (HttpResponse) -> NetworkResponse<BodyType>? = { null },
+    performRequest: () -> HttpResponse
+): NetworkResponse<BodyType> = handlingNetworkException {
+    val result = performRequest()
+    val status = result.status
+    return if (status.isSuccess()) {
+        NetworkResponse.Success(result.body(), result)
+    } else {
+        unsuccessfulResponseOverride(result) ?: when (status.value) {
+            HttpStatusCode.Unauthorized.value -> {
+                kaliumLogger.e("Unauthorized request, $result")
+                NetworkResponse.Error(KaliumException.Unauthorized(status.value))
             }
+            in (300..399) -> NetworkResponse.Error(KaliumException.RedirectError(result.body()))
+            in (400..499) -> NetworkResponse.Error(KaliumException.InvalidRequestError(result.body()))
+            in (500..599) -> NetworkResponse.Error(KaliumException.ServerError(result.body()))
```

Comment:
It will be nice if we can try catch each time there is a response serialization `result.body()` since BE is not consistent with the error response and can response with empty/html body and in result the the  serialization will throw an exception.